### ISSUE TITLE
Incorrect escaping in Asis output in github document

### ISSUE THREAD

**Issue by eitsupi** [rstudio/rmarkdown#2479]
From https://github.com/pola-rs/r-polars/pull/125#issuecomment-1518933933

I defined and used the following function to write out an HTML table and noticed that `[` and `]` were improperly escaped.

Note that this mimics the HTML output of the Python polars package, and the polars HTML tables output from the python code chunks do not have this problem.

````rmd
---
title: test
output:
  github_document:
    df_print: paged
---

```{r}
#| include: false
to_html_table <- function(x, max_cols = 75, max_rows = 40) {
  escape_html <- getFromNamespace("escape_html", "knitr")
  omit_chr <- "&hellip;"

  .dim <- dim(x)

  df_height <- .dim[1]
  df_width <- .dim[2]

  row_idx <- .idx(max_rows, df_height)
  col_idx <- .idx(max_cols, df_width)

  cols <- names(x)[col_idx]
  col_names <- cols |>
    escape_html()
  col_types <- x |>
    .get_dtype_strings() |>
    escape_html()

  if (max_cols <= df_width) {
    col_names <- .cut_off(col_names, max_cols, omit_chr)
    col_types <- .cut_off(col_types, max_cols, omit_chr)
  }

  .header_names <- col_names |>
    .tag("th") |>
    .tag("tr")

  .header_dtypes <- col_types |>
    .tag("td") |>
    .tag("tr")

  .header_all <- .header_names |>
    paste0(.header_dtypes) |>
    .tag("thead")

  .env_str_len <- Sys.getenv("POLARS_FMT_STR_LEN")
  .str_len <- ifelse(.env_str_len == "", 15, as.integer(.env_str_len))

  chr_mat <- sapply(cols, \(col) as.character(x[row_idx, col, drop = TRUE], str_length = .str_len)) |>
    escape_html() |>
    matrix(nrow = length(row_idx))

  if (max_cols <= df_width) {
    .seq <- seq_along(cols)
    chr_mat <- cbind(
      chr_mat[, head(.seq, max_cols %/% 2)],
      omit_chr,
      chr_mat[, tail(.seq, max_cols %/% 2)]
    )
  }

  if (max_rows <= df_height) {
    .seq <- seq_along(row_idx)
    chr_mat <- rbind(
      chr_mat[head(.seq, max_rows %/% 2), ],
      omit_chr,
      chr_mat[tail(.seq, max_rows %/% 2), ]
    )
  }

  .body <- chr_mat |>
    t() |>
    as.data.frame() |>
    sapply(\(x) .tag(x, "td")) |>
    .tag("tr") |>
    .tag("tbody")

  .style <- "<style>
.dataframe > thead > tr > th,
.dataframe > tbody > tr > td {
  text-align: right;
}
</style>
"

  .shape <- sprintf("<small>shape: (%s, %s)</small>", .dim[1], .dim[2])

  paste0(.header_all, .body) |>
    .tag("table", c(border = "1", class = "dataframe")) |>
    (\(x) (paste0(.style, .shape, x)))() |>
    .tag("div")
}

.idx <- function(.max, .length) {
  if (.max <= .length) {
    out <- c(seq_len(.max %/% 2), seq(.length - .max %/% 2 + 1L, .length))
  } else {
    out <- seq_len(.length)
  }
  out
}

.cut_off <- function(.vec, .max, omit_chr) {
  c(head(.vec, .max %/% 2), omit_chr, tail(.vec, .max %/% 2))
}

#' @param .elements chr vector
#' @param .tag single chr
#' @param .attr named chr vector
#' @return single charactor
#' @examples
#' .tag(letters[1:2], "tr")
#' @noRd
.tag <- function(.elements, .tag, .attr = NULL) {
  if (!is.null(.attr)) {
    .pre <- paste0("<", .tag, " ", paste0(sprintf('%s="%s"', names(.attr), .attr), collapse = " "), ">")
  } else {
    .pre <- c(paste0("<", .tag, ">"))
  }

  .post <- paste0("</", .tag, ">")

  paste0(.pre, .elements, .post, collapse = "")
}

.get_dtype_strings <- function(df) {
  if (inherits(df, "DataFrame")) {
    df$dtype_strings()
  } else {
    sapply(names(df), \(x) pillar::type_sum(df[, x, drop = TRUE])) |>
      unname()
  }
}
```

```{r}
#| label: test
#| results: asis
data.frame(a = "foo [bar]") |> to_html_table() |> writeLines()
```

```{r}
#| label: test
```
````

The output md file is like this.

````md
test
================

``` r
data.frame(a = "foo [bar]") |> to_html_table() |> writeLines()
```

<div>

<style>
.dataframe > thead > tr > th,
.dataframe > tbody > tr > td {
  text-align: right;
}
</style>
<small>shape: (1, 1)</small>
<table border="1" class="dataframe">
<thead>
<tr>
<th>
a
</th>
</tr>
<tr>
<td>
chr
</td>
</tr>
</thead>
<tbody>
<tr>
<td>
foo \[bar\]
</td>
</tr>
</tbody>
</table>

</div>

``` r
data.frame(a = "foo [bar]") |> to_html_table() |> writeLines()
```

    ## <div><style>
    ## .dataframe > thead > tr > th,
    ## .dataframe > tbody > tr > td {
    ##   text-align: right;
    ## }
    ## </style>
    ## <small>shape: (1, 1)</small><table border="1" class="dataframe"><thead><tr><th>a</th></tr><tr><td>chr</td></tr></thead><tbody><tr><td>foo [bar]</td></tr></tbody></table></div>
````

The environment is as follows

```r
> xfun::session_info('rmarkdown')
R version 4.2.2 (2022-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.1 LTS, RStudio 0

Locale:
  LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
  LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

Package version:
  base64enc_0.1.3   bslib_0.4.2       cachem_1.0.6      cli_3.5.0         digest_0.6.31     ellipsis_0.3.2    evaluate_0.19     fastmap_1.1.0     fontawesome_0.5.1
  fs_1.5.2          glue_1.6.2        graphics_4.2.2    grDevices_4.2.2   highr_0.9         htmltools_0.5.4   jquerylib_0.1.4   jsonlite_1.8.4    knitr_1.41       
  lifecycle_1.0.3   magrittr_2.0.3    memoise_2.0.1     methods_4.2.2     mime_0.12         R6_2.5.1          rappdirs_0.3.3    rlang_1.0.6       rmarkdown_2.21.3 
  sass_0.4.4        stats_4.2.2       stringi_1.7.8     stringr_1.5.0     tinytex_0.43      tools_4.2.2       utils_4.2.2       vctrs_0.5.1       xfun_0.35        
  yaml_2.3.6       

Pandoc version: 2.19.2
```

## Checklist

When filing a _bug report_, please check the boxes below to confirm that you have provided us with the information we need. Have you:

- [x] [formatted your issue](https://yihui.org/issue/#please-format-your-issue-correctly) so it is easier for us to read?

- [x] included a minimal, self-contained, and reproducible example?

- [x] pasted the output from `xfun::session_info('rmarkdown')` in your issue?

- [ ] upgraded all your packages to their latest versions (including your versions of R, the RStudio IDE, and relevant R packages)?

- [x] installed and tested your bug with the development version of the rmarkdown package using `remotes::install_github("rstudio/rmarkdown")`?

**Comment by yihui** [comment]
I'm afraid this is the behavior of Pandoc and perhaps we can't do anything about it...

```sh
echo 'foo [bar]' | pandoc -f markdown -t gfm
```
```
foo \[bar\]
```

**Comment by eitsupi** [comment]
Thanks for giving me some context for this!

However, in the output of python code chunks using the reticulate's Python engine, which outputs nearly identical HTML, `[` and `]` does not appear to be escaped.

In the image I noticed this (https://github.com/pola-rs/r-polars/pull/125#issuecomment-1518933933), comparing the output by Python above with the output by R below, only the bottom has the rightmost column name escaped.

![image](https://user-images.githubusercontent.com/50911393/233817692-a7398e28-cd2a-43d5-b2e1-84aec730f5ad.png)

I'm wondering if the R engine can work around this too.

**Comment by yihui** [comment]
If the output is actually raw HTML, the best way to preserve it (i.e., avoid any escaping) is to put it in a raw HTML block:

````md
```{=html}

```
````

I'm not sure if that's what the Python code does.

You can wrap output in `knitr::raw_html()` to generate a raw HTML block.

**Comment by eitsupi** [comment]
Thank, reticulate seems using `knitr::raw_html()`, so I will try that.
https://github.com/rstudio/reticulate/blob/63a5c3ef229a5ac76bb299575a3c36e6ae650b85/R/knitr-engine.R#L614-L618

**Comment by eitsupi** [comment]
`knitr::raw_html()` works fine for me, thanks!

**Comment by cderv** [comment]
For referenc, pandoc answer on this: https://github.com/jgm/pandoc/issues/4830 

TL;DR: This is by design to prevent unintentional links.

Escaping using raw block is probably the best here.
